ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogCache.java

```diff
@@ -342,10 +342,10 @@ int getMaxCachedSegments() {
   }
 
   void loadSegment(LogPathAndIndex pi, boolean keepEntryInCache,
-      Consumer<LogEntryProto> logConsumer) throws IOException {
+      Consumer<LogEntryProto> logConsumer, long lastIndexInSnapshot) throws IOException {
     LogSegment logSegment = LogSegment.loadSegment(storage, pi.getPath().toFile(),
```

Comment:
This is for loading log segments from disk. Is it possible that a log segment from disk has INVALID_LOG_INDEX as the endIndex?

Reply:
Yes it is the case for open segment.

Reply:
The endIndex would be updated during LogSegment.loadSegment, even for openSegments.  So if the endIndex is still INVALID_LOG_INDEX, we should not load it, right?

Reply:
For open segment endIndex is INVALID_LOG_INDEX and we will load it. We use LogSegment#newOpenSegment to create an open segment.
```  
SegmentedRaftLogInputStream(File log, long startIndex, long endIndex, boolean isOpen,
                                      RaftLogMetrics raftLogMetrics) {
    if (isOpen) {
      Preconditions.assertTrue(endIndex == INVALID_LOG_INDEX);
    } else {
      Preconditions.assertTrue(endIndex >= startIndex);
    }
...
```

Reply:
For open segment endIndex is INVALID_LOG_INDEX and we will load it. We use LogSegment#newOpenSegment to create an open segment.
```  
SegmentedRaftLogInputStream(File log, long startIndex, long endIndex, boolean isOpen,
                                      RaftLogMetrics raftLogMetrics) {
    if (isOpen) {
      Preconditions.assertTrue(endIndex == INVALID_LOG_INDEX);
    } else {
      Preconditions.assertTrue(endIndex >= startIndex);
    }
...
```

Reply:
Here is the flow as I understand. Please correct if I am wrong.
1. New open segment is created in LogSegement#loadSegment()
```
final LogSegment segment = isOpen ?
        LogSegment.newOpenSegment(storage, start, raftLogMetrics) :
        LogSegment.newCloseSegment(storage, start, end, raftLogMetrics);
```
2. The segment file is read from disk and each entry is appended to this newly created open LogSegment.
```
final int entryCount = readSegmentFile(file, start, end, isOpen, corruptionPolicy, 
   raftLogMetrics, entry -> {
      segment.append(keepEntryInCache || isOpen, entry);
      if (logConsumer != null) {
        logConsumer.accept(entry);
      }
    });
```
3. For each entry we do LogSegment.append() which updates the endIndex of the open segment to the entry's index. 
```
private void append(boolean keepEntryInCache, LogEntryProto entry) {
  ...
  ...
  totalSize += getEntrySize(entry);
  endIndex = entry.getIndex();
}
```
4. When LogSegment.loadSegment() call is completed in SegmentedRaftLogCache#loadSegment(), the open segment's endIndex is the index of the last entry in that segment. If the open segment has only entries less than the lastIndexInSnapshot, we should not add it to cache.
```
void loadSegment(LogPathAndIndex pi, boolean keepEntryInCache,
      Consumer<LogEntryProto> logConsumer, long lastIndexInSnapshot) throws IOException {
    LogSegment logSegment = LogSegment.loadSegment(storage, pi.getPath().toFile(),
        pi.getStartIndex(), pi.getEndIndex(), pi.isOpen(), keepEntryInCache, logConsumer, raftLogMetrics);
    if (logSegment != null && logSegment.getEndIndex() > lastIndexInSnapshot) {
      addSegment(logSegment);
    }
  }
```

Reply:
I misunderstood the previous comment. I agree. It's better to make the check after loading the segment. I was talking about the optimisation to check LogPathAndIndex in loadSegment but it may be an overkill.
```
void loadSegment(LogPathAndIndex pi, boolean keepEntryInCache,
      Consumer<LogEntryProto> logConsumer, long lastIndexInSnapshot) throws IOException {
    if (pi.getEndIndex() > lastIndexInSnapshot || pi.isOpen()) {
      LogSegment logSegment = LogSegment.loadSegment(storage, pi.getPath().toFile(), 
          pi.getStartIndex(), pi.getEndIndex(), pi.isOpen(), keepEntryInCache, logConsumer, raftLogMetrics);
      if (logSegment != null && logSegment.getEndIndex() > lastIndexInSnapshot) {
        addSegment(logSegment);
      }
    }
  }
```